README
$Id: README,v 1.11 2002/10/01 22:32:27 rowan Exp $

This is exhaust version 1.9, a memory array redcode simulator.  It is
mostly in the public domain (see the file COPYING for details).

Files:
        changelog          Log of Change.
        COPYING            lack of license.
        Makefile           file for make.
        README             This file.
	NEWS		   Notes about things that might break.
        asm.c              (Dis)Assembler.
        asm.h              Token codes and prototypes for assembler.
        exhaust.c          Simplified pMARS-like simulator.
        exhaust.h          Global constants, structures and types.
        insn.h             Instruction encoding specs.
        sim.c              Simulator proper.
        sim.h              Simulator prototypes.
	pspace.c	   P-space implementation.
	pspace.h	   P-space prototypes.
        test.pl            Perl program to test for differences between
                           simulators.
        t/*.red		   Redcode for tests.
	t/*.rc		   Load files for tests.

To compile on UNIX or on DOS with djgpp first edit the makefile to
your comfort and then run `make' to get an executable `exhaust'.  If
you have gcc you hopefully don't need to make any large makefile
modifications.  See the section `How to use the simulator and friends'
below for an overview of how to use the simulator in your programs.

Thanks to the pMARS authors and Ken Espiritu for ideas implemented in
this simulator.  Especially Ken's effective addressing calculation code
from pMARS 0.8.6 was adapted for a great speed-up.


* How to use the simulator and friends
--------------------------------------

This is a quick run-down on how to use the simulator and assembler
with your program.  Special considerations for P-Space are detailed in
the section about p-space below, so here non-P warriors are assumed.
After munging your source file, remember to link your executable with
sim.o, pspace.o, and asm.o!  You don't need asm.o if you don't use the
assembler or disassembler.

The main things that you should do are:
  - generate (perhaps by assembling) the warrior code
  - allocate the internal buffers
  - clear core and load the warriors into it
  - call the simulator to do it's thing

In the examples, I assume you use a warrior structure similar to
exhaust.h's struct warrior_st, but it is not canonical in any way.
Here I rely on the following fields:

        insn_t code[ MAXLENGTH ];   /* code of warrior */
        int len;                    /* length of warrior code */
        int start;                  /* start relative to first insn */

* The assembler: 

i) You need some declarations and includes:

#include "exhaust.h"
#include "asm.h"

        warrior_t w1, w2;       /* warrior structs hold code and misc. info */
        int CORESIZE = 8000,    /* size of core */
            PROCESSES = 8000,   /* max. number of processes per warrior */
            NWARRIORS = 2,      /* number of warriors */
            CYCLES = 80000;     /* cycles until tie */

ii) Then you want to assemble your warriors, say aeka and Rave:

        asm_fname( "aeka.rc", &w1, CORESIZE );
        asm_fname( "rave.rc", &w2, CORESIZE );

You could also use asm_file(FILE *, warrior_t *, CORESIZE) to assemble
from a stream.


* Simulator:

i) The simulator interface is simple.  First you need to allocate and
query the address of the core memory:

#include "exhaust.h"
#include "sim.h"
        insn_t *core;

        core = sim_alloc_bufs( NWARRIORS, CORESIZE, PROCESSES, CYCLES );

If any of the allocations failed the returned value will be NULL.  You
should check for that. When you're done with the core memory and other
buffers, call sim_free_bufs(void), but note that sim_alloc_bufs()
frees any memory it previously allocated.  i.e. you can successively
call it without intervening calls to sim_free_bufs().

For backward compatibility you can also use use sim() as follows to
allocate with the default CORESIZE=PROCESSES=8000:

        (void) sim( -1, 0, 0, CYCLES, &core );

ii) Then you clear core and load the warriors:

        memset( core, 0, sizeof(insn_t) * CORESIZE );
or
	sim_clear_core();

and
	sim_load_warrior(0,   &(w1.code[0]), w1.len );
	sim_load_warrior(pos, &(w2.code[0]), w2.len );

where `pos' is the core address where you want to load w2.  This
differs slightly from the previous release where you had to:

        memcpy( core,       w1.code, sizeof(insn_t) * w1.len );
        memcpy( core + pos, w2.code, sizeof(insn_t) * w2.len );

You can still do that.  (Note however, that the flags field of
instructions should be cleared as the simulator won't do it anymore.
sim_load_warrior() does this for you.  You can get the old behaviour
by defining the preprocessor variable SIM_STRIP_FLAGS to 1 at compile
time.)


iii) Right, now you're ready to fight.  If you have only one or two
warriors, you can use sim():

        result = sim( 2, w1.start, (w2.start+pos)%CORESIZE, CYCLES, NULL);

The result is 0 if w1 won, 1 if w2 won, and 2 on a tie (-1 if the
simulator panics).

For multi-warrior battles with more than two warriors you need
some extra declarations:

        field_t war_pos_tab[ NWAR ];    /* a table of core addresses
                                         * that specify where the warriors
                                         * start their execution. */
        int death_tab[ NWAR ];          /* the simulator stores the
                                         * results here. (see below) */

Then allocate the buffers and obtain the core address using
sim_alloc_bufs(), clear the core, and load the warriors into it as
before.  Save the positions where you loaded the warriors into
war_pos_tab[].  Now you're ready to fight, so call sim_mw():

        int alive_cnt;
        alive_cnt = sim_mw( NWAR, war_pos_tab, death_tab );

The return value is the number of warriors alive at the end of the
round.  Dead warriors are recorded into death_tab[] by order of death.
i.e. death_tab[0] contains the index of the first warrior to die,
death_tab[1] the index of the second to go, and so on.


* The disassembler:

If you want to disassemble an instruction use dis1() or discore() from
asm.c.  Examples:

#include "exhaust.h"
#include "asm.h"

        char str[60];
        insn_t instr;

        instr.in = _OP( SPL, mF, DIRECT, BPREDEC );
        instr.a = instr.b = 0;
        core[1234] = instr;

        dis1( str, core[1234], CORESIZE ); printf("%s\n", str);
        dis1( str, instr, CORESIZE );      printf("%s\n", str);

                                        /* disassemble addresses 5..14 */
        discore( core, 5, 15, CORESIZE);/* and print to stdout. */


* P-Space
---------

The simulator interface is oriented toward playing single battles, but
p-space by its nature requires a multi-round interface.  Hence there
are some complications when playing p-space warriors.  The main change
for P-spacer battles over normal ones is that you need to pair
p-spaces to warriors.  Also, if you are running multiple battles you
need to reinitialise the p-spaces between each battle.

The basic model is simple: the simulator maintains an array of
NWARRIORS p-spaces that you may manipulate.  The ith p-space in this
array always corresponds to the ith warrior passed to sim() or
sim_mw(), so that the first warrior always accesses the first p-space,
the second the second, and so on.  The pairing betweens p-spaces and
warriors is up to you.  After each round the p-space location 0 of
each warrior is updated with either a) zero if the warrior died or b)
the number of warriors alive at the end of the battle.

[A note about naming conventions: pspace_XXX() functions are mostly
local in nature -- they refer only to their argument p-spaces.  The
sim_XXX_pspace() functions are about the collection of p-spaces
maintained by the simulator.]


* Accessing p-spaces

To get a pointer to the ith p-space you use the function
sim_get_pspace():

	pspace_t *p = sim_get_pspace(i);

Then to access the cells of `p', use the functions pspace_get() and
pspace_set():

	contents_of_cell_123 = pspace_get(p, 123);

	pspace_set(p, 123, new_contents);


* P-space initialisation; resets

 - The size of p-space is specified at memory allocation time using
   the function sim_alloc_bufs2().  It is exactly like
   sim_alloc_bufs(), except it takes as an additional parameter the
   size of pspace, which must be positive.  The later function
   defaults p-space to CORESIZE/16 cells, so you don't need to change
   it.

 - The function sim_clear_pspaces() clears the p-spaces of all warriors
   and stores CORESIZE-1 into their p-space location 0.  P-Space is
   cleared on allocation, so you only need to call this function
   between battles, but not rounds.  This function doesn't change
   the sharing status of p-spaces.

 - To completely reinitialise p-spaces and to make p-spaces of each
   warrior private again, call the sim_reset_pspaces() function.  This
   also calls sim_clear_pspaces().


* Pairing p-spaces to warriors

If you are changing the order in which warriors execute within each
cycle on a per-round basis, you need to order p-spaces as well.  For
example, pMARS cyclically permutes the order of warriors so that every
even round the first warrior executes first in each cycle, and every
odd round the second warrior executes first.  With non-pspace warriors
you only need to give the starting positions of the warriors in the
`war_pos_tab[]' argument in the order you want the warriors to
execute, but with p-warriors involved you need to match p-spaces to
warriors as well.

 - The simulator maintains an array of pointers to p-space structures,
   one for each warrior.  The ith p-space structure corresponds to the
   ith warrior in the `war_pos_tab[]' array argument of warrior
   starting positions passed to sim_mw().  Similarly for the two
   warriors given to the sim() function.  You need to order this array
   of pointers to match the order in which you are passing warrior
   starting positions to the simulator.

   To access the array, use sim_get_pspaces():

	pspace_t **pspaces;
	pspaces = sim_get_pspaces();

	// now permute the pointers in the pspaces array into the
	// same order as your warriors.
	...

   Of course, if you are not changing the order in which warriors
   execute on a per-round basis, you don't need to permute the pspaces
   array.  This may be suitable in some cases.


* Sharing p-spaces

There is only minimal support for PINs and p-space sharing in that the
assembler understands the `PIN' pseudo-op, but you must match PINs to
p-spaces yourself.

  - The warrior structure warrior_st in exhaust.h needs the extra fields
  
	int have_pin;			// boolean: is field `pin' valid?
	u32_t pin;			// PIN of warrior, possibly.

    that are used when assembling from files/streams.  The interface
    to asm_line() has changed accordingly so it can identify and
    return the PIN pseudo-op.

  - To share p-spaces of two warriors, you need to call the function
    pspace_share(p1, p2) for two pointers p1, p2 to p-spaces.  The
    contents of p-space p1 becomes the contents of the shared p-space.
    P-space location 0 is never shared.

  - The function pspace_privatise(p) resets the p-space p to be
    private.  The contents of p1 are then undefined.



* Files that the assembler eats
-------------------------------

In short: the load files that the '94 draft specifies are accepted.
Specifically, the files may contain only `;' comments, white space,
instructions, and pseudo-ops.  The instructions must follow this form:

  [opt START label]    OPCODE.MODIFIER  MODE integer , MODE integer

Various idiosyncrasies:

 - You can't implicitly specify direct addressing mode by dropping `$'.

 - The start of a warrior is determined by the last seen appearance of
   the START label or `ORG'.  Of the pseudo-ops using `END label' or
   `ORG label' aren't supported; only `ORG integer' is.

 - `END' stops warrior assembly.  You still can assemble from the
   same file or stream another warrior that follows though.

 - special KotH comments aren't recognised.

To support using pMARS as an assembler any line that begins with
`Program' is ignored.  So you could first say

  pmars -r 0 Fancy_Source.red >Simple_Source.rc

and then use the asm.c assembler to import Simple_Source.rc into your
program.  Note that pmars doesn't output PINs of warriors.


* When the simulator crashes
----------------------------

... or is just plain wrong by producing incorrect results please tell
me about it at the address given below.  Ditto for the assembler or
anything else in these files.  I'd appreciate a simple test case that
illustrates the bug if possible.

The author can be reached by email at: jpihlaja@cc.helsinki.fi


NEWS
* Version 1.9.2:

- Lifted hardcoded upper limit on the number of warriors.

* Version 1.9:

- Flags in instructions aren't masked when decoding instructions
  during simulation anymore.  You should either clear the flags fields
  before loading, or use the new function sim_load_warrior() to load
  your warriors.  If that's not feasible, define the preprocessor
  variable SIM_STRIP_FLAGS to be 0 when compiling.

  As far as I can tell nobody was using the flags field anyway.  Sorry
  if this breaks anyones code.

- P-space!  See the README for details.

- You need to link the file pspace.o whenever you are linking sim.o
  regardless of whether p-space is in use or not.

- To accomodate p-space the encoding of instructions has changed.  One
  flag bit has evaporated to make room for LDP/STP, and the order of
  enumeration of opcodes differs from version 1.7.  If you've been
  using the macros in insn.h or the assembler, then there's nothing to
  worry about.  Sorry if this breaks anyones code.

- Now placed into the Public Domain.

- asm.h/asm_line() can now return -3 == ASMLINE_PIN to identify the
  PIN pseudo-op.

- The warrior_st struct in exhaust.h has new fields `have_pin' and `pin'.

- Other changes to fix some compiler warnings/sign oddities.  Some
  arguments of public prototypes have changed in signedness.

- No more ABORT_TIES_EARLY, but that's OK as 1.8 was never released.

- The example program exhaust.c now supports multiwarrior battles and
  p-space.  Slightly cleaned up as well.

* Version 1.8:

- New compile-time option ABORT_TIES_EARLY to detect unwinnable
  configurations where there are no deadly instructions in core.


asm.c
/* asm.c: primitive redcode assembler
 * $Id: asm.c,v 1.8 2002/10/01 22:24:46 rowan Exp $
 */

/* This file is part of `exhaust', a memory array redcode simulator.
 * Author: M Joonas Pihlaja
 * Public Domain.
 */

/* The format of lines with instructions should be:
 *
 * [START]	OPCODE.MODIFIER   A-MODE INT , B-MODE INT
 *
 * The ORG pseudo-op is ignored, as is the label after an optional
 * END (if given). The only label recognised is START.  No fuss over
 * the amount of white space, as long as it exists where required.
 * 
 * Comments are recognised and discarded as is any line starting
 * with "Program".  The output from `pmars -r 0 Your_Real_Source.red'
 * should assemble fine with this tiny assembler.
 *
 *
 * Functions in this file:
 *     
 *     asm_line(), asm_file(), asm_fname(), dis1(),
 *     discore()
 */
#include <stdio.h>
#ifdef SYSV
#include <strings.h>
#else
#include <string.h>
#endif
#include <stdlib.h>
#include <ctype.h>

#include "exhaust.h"
#include "insn.h"
#include "asm.h"

/* str_tok_t: container for tokens we identify.
 */
typedef struct str_toks_st {
  char *s;			/* name of token */
  int c;			/* token code */
} str_toks_t;

/* Data
 *
 * tok_buf[]: globally used to keep the contents of string tokens
 * tok_int:   if the token was a TOK_INT, the value of the token is here
 *
 * str_toks[]: table of multicharacter tokens we identify
 *
 */

#define MAX_ALL_CHARS 256
static char tok_buf[MAX_ALL_CHARS];
static int tok_int;

static str_toks_t str_toks[] = {
    { "DAT", TOK_DAT },		/* opcodes */
    { "SPL", TOK_SPL },
    { "MOV", TOK_MOV },
    { "DJN", TOK_DJN },
    { "ADD", TOK_ADD },
    { "JMZ", TOK_JMZ },
    { "SUB", TOK_SUB },
    { "MOD", TOK_MOD },
    { "CMP", TOK_SEQ },
    { "SEQ", TOK_SEQ },
    { "JMP", TOK_JMP },
    { "JMN", TOK_JMN },
    { "SNE", TOK_SNE },
    { "MUL", TOK_MUL },
    { "DIV", TOK_DIV },
    { "SLT", TOK_SLT },
    { "NOP", TOK_NOP },
    { "LDP", TOK_LDP },
    { "STP", TOK_STP },

    { "ORG", TOK_ORG },		/* pseudo-ops */
    { "END", TOK_END },
    { "PIN", TOK_PIN },
    { "START", TOK_START },

    { "F", TOK_mF },		/* modifiers */
    { "A", TOK_mA },
    { "B", TOK_mB },
    { "AB", TOK_mAB },
    { "BA", TOK_mBA },
    { "X", TOK_mX },
    { "I", TOK_mI },
    { NULL, 0 }			/* sentinel */
};



/* NAME
 *     get_tok -- read the next token from a string
 * 
 * SYNOPSIS
 *     const char *get_tok( const char *s, int *tok );
 * 
 * INPUTS
 *     s -- string to read token from
 *     tok -- where we store the token code of the read token
 * 
 * RESULTS
 *     The token code of the read token is stored into *tok,
 *     with 0 signifying end of input.
 *
 *     If the token was an integer, its value is stored into
 *     the global `tok_int'.  Integers may be in any base >= 10
 *     as according to strtol().
 *
 *     String tokens are converted to upper case when storing
 *     them into the global `tok_str[]'.  They are concatenated
 *     at 255 characters.
 * 
 * RETURN VALUE
 *      Pointer to the character past the read token, or 
 *	to the nul character if at end of input.
 * 
 * GLOBALS
 *     tok_buf[]    -- a string or char token is copied here
 *     tok_int      -- the value of an integer token
 *     str_toks[]   -- used to identify string tokens
 */

/* skip_white(): returns ptr. to next non-whitespace char in s */
static
const char *
skip_white(const char *s)
{
  while ( isspace(*s) ) s++;
  return s;
}

static
const char *
get_tok( const char *s, int *tok )
{
  char *tok_str = tok_buf;
  int i;

  s = skip_white(s);
  if ( *s == 0 )    return (*tok = 0, s);

  /*
   * Tokenize strings.
   *
   * String tokens must start with a letter and consist of
   * letters, digits, and underscores.  Strings are
   * converted to upper case.
   */
  tok_buf[1] = tok_buf[0] = 0;

  i = 0;
  if ( isalpha(*s) )
    while ( (isalnum(*s) || *s == '_') && ++i < MAX_ALL_CHARS )
      *tok_str++ = toupper(*s++);
  *tok_str = 0;

  if ( tok_str > tok_buf ) {
    /*
     * was a string token -- identify it by searching through
     * the str_toks[] array.
     */
    for ( i = 0; str_toks[i].s ; i++ ) {
      if ( 0 == strcmp( str_toks[i].s, tok_buf ) ) {
	*tok = str_toks[i].c;
	return s;
      }
    }
    *tok = TOK_STR;		/* normal string, not special */
    return s;
  }


  /*
   * Tokenize ints.
   * Must match /-?[0-9]/
   */
  if ( isdigit(*s) ||  ( *s == '-' && isdigit(*(s+1)) )) {
    char *endptr;
    tok_int = strtol( s, &endptr, 0 );
    *tok = TOK_INT;
    return endptr;
  }


  /*
   * Tokenize addressing modes and pass single chars
   */

  tok_buf[0] = *s;		/* store char value as single */
  tok_buf[1] = 0;		/* char string. */

  switch ( *tok = *s++ ) {
  case '$': *tok = TOK_DIRECT;          break;
  case '#': *tok = TOK_IMMEDIATE;	break;
  case '*': *tok = TOK_AINDIRECT;	break;
  case '@': *tok = TOK_BINDIRECT;	break;
  case '{': *tok = TOK_APREDEC;		break;
  case '<': *tok = TOK_BPREDEC;		break;
  case '}': *tok = TOK_APOSTINC;	break;
  case '>': *tok = TOK_BPOSTINC;	break;
  }

  return s;
}



/* NAME
 *     panic_bad_token -- issue an error message for a bad token and exit(1)
 * 
 * SYNOPSIS
 *     void panic_bad_token( int tok, const char *expected );
 * 
 * INPUTS
 *     tok -- token code of unexpected token
 *     expected -- a string describing what kind of token
 *		   was expected.  e.g. "a modifier".
 * 
 * RESULTS
 *     A message Informing the user of the unexpected token,
 *     its possible semantic value, and what type of token
 *     was expected instead.
 * 
 * GLOBALS
 *     tok_buf, tok_int -- if the token has semantic value we look
 *                         for it here.
 * BUGS
 *     The error message should be much better -- not even location
 *     in the source is given here. *sigh*
 */
static
void
panic_bad_token( int tok, const char *expected ) 
{
  char *errstr = NULL;
  char buf[30];

  memset(buf, 0, 30);

  /* make an errstr
   */
  if ( tok_buf[0] )
    errstr = tok_buf;
  if ( tok == TOK_INT ) {
    sprintf(buf, "%d", tok_int );
    errstr = buf;
  }

  /* complain and exit with error code
   */
  fprintf(stderr, "token '%s' not %s\n", errstr, expected );
  exit(1);
}


/* NAME
 *     asm_line -- assemble a line to an instruction
 * 
 * SYNOPSIS
 *     int asm_line( const char *line, insn_t *in, unsigned int CORESIZE );
 * 
 * INPUTS
 *     line -- line to assemble
 *     in   -- instruction to assemble into
 *     CORESIZE -- size of core
 * 
 * RESULTS
 *     If there was anything to assemble, it is assembled into
 *     `in'.  If there was a START label, the corresponding flag
 *     is set in the instructions flags.  Incomplete or erroneous
 *     input prompt a quick error message and exit(1).
 *
 *     If the 'ORG start-address' construct is encountered where
 *     `start-address' is an integer, then the `in->a' field contains
 *     the offset in instructions from the start of the warrior
 *     where the warrior should start execution.
 *
 *     If 'PIN id' is encountered, where `id' is an integer, then the
 *     `in->a' field contains the `id'.
 * 
 * RETURN VALUE
 *     ASMLINE_PIN  : pseudo-op 'PIN' encountered, id saved in `in->a'.
 *     ASMLINE_ORG  : pseudo-op 'ORG' encountered, warrior start
 *                    saved in `in->a'.
 *     ASMLINE_DONE : done assembling, END opcode found, nothing assembled.
 *     ASMLINE_NONE : nothing to assemble on this line.
 *     ASMLINE_OK   : assembled instruction into `in' OK.
 * 
 * GLOBALS
 *     tok_int, tok_buf[], str_toks[] somewhere down the line.
 */

int
asm_line( const char *line, insn_t *in, unsigned int CORESIZE  )
{
  const char *s = line;
  int tok;
  int flags = 0;
  int op, m, ma, mb;		/* opcode, modifier, a-mode, b-mode */

  s = get_tok( s, &tok );
  if ( tok == 0 ) return ASMLINE_NONE;

  /*
   * Ignore string lines '^Program.*' and comments.
   */
  if ( tok == TOK_STR && 0 == strcmp( "PROGRAM", tok_buf ))
  {
    return ASMLINE_NONE;
  }
  if ( tok == ';' ) return ASMLINE_NONE;

  /*
   * Now match the instruction's various components:
   *   [START label,] opcode, modifier, a-mode, a-value, b-mode, b-value
   */

  /* Match possible start label
   */
  if ( tok == TOK_START ) {
    flags |= fl_START;
    s = get_tok( s, &tok );
  }

  /* Match opcode
   */
  if ( is_tok_pseudoop(tok) ) {
    switch ( tok ) {
    case TOK_END:
      return ASMLINE_DONE;	/* signal done assembling */

    case TOK_ORG:
      s = get_tok( s, &tok );	/* get the next token */

      if ( tok == TOK_START )	/* ignore: */
	return ASMLINE_NONE;	/* start label already matched and processed */

      if ( tok != TOK_INT ) {
	panic_bad_token( tok, "an integer -- an int or \"START\" "
			 "follows ORG" );
      }
      in->a = tok_int;
      return ASMLINE_ORG;

    case TOK_PIN:
      s = get_tok( s, &tok );
      if ( tok != TOK_INT ) {
	panic_bad_token( tok, "an integer -- PIN must be an unsigned integer");
      }
      in->a = tok_int;
      return ASMLINE_PIN;

    default:
      panic_bad_token( tok, "a pseudo-op (internal assembler error)" );
    }
  }
  if (!( is_tok_opcode(tok)))
    panic_bad_token( tok, "an opcode" );

  op = DAT;
  switch(tok) {
  case TOK_DAT: op = DAT; break;
  case TOK_SPL: op = SPL; break;
  case TOK_MOV: op = MOV; break;
  case TOK_JMP: op = JMP; break;
  case TOK_JMZ: op = JMZ; break;
  case TOK_JMN: op = JMN; break;
  case TOK_ADD: op = ADD; break;
  case TOK_SUB: op = SUB; break;
  case TOK_SEQ: op = SEQ; break;
  case TOK_SNE: op = SNE; break;
  case TOK_MUL: op = MUL; break;
  case TOK_DIV: op = DIV; break;
  case TOK_DJN: op = DJN; break;
  case TOK_SLT: op = SLT; break;
  case TOK_MOD: op = MODM; break;
  case TOK_NOP: op = NOP; break;
  case TOK_LDP: op = LDP; break;
  case TOK_STP: op = STP; break;
  default:
    panic_bad_token( tok, "an opcode" );
  }

  /* Match modifier
   */
  s = get_tok( s, &tok );	/* first the '.' */
  if ( tok != '.' )
    panic_bad_token( tok, "'.'" );

  s = get_tok( s, &tok );	/* then the modifier itself */
  if ( ! is_tok_modifier(tok) )
    panic_bad_token( tok, "a modifier");
  m = tok - TOK_mF;

  /* Match a-field addressing mode and a-field
   */
  s = get_tok( s, &tok );
  if ( ! is_tok_mode(tok) )
    panic_bad_token( tok, "an addressing mode specifier");
  ma = tok - TOK_DIRECT;

  s = get_tok( s, &tok );
  if ( tok != TOK_INT )
    panic_bad_token( tok, "an integer");
  in->a = MODS(tok_int,CORESIZE);

  /* Match comma
   */
  s = get_tok( s, &tok );
  if ( tok != ',' )
    panic_bad_token( tok, "','" );

  /* Match b-field addressing mode and a-field
   */
  s = get_tok( s, &tok );
  if ( ! is_tok_mode(tok) )
    panic_bad_token( tok, "an addressing mode specifier");
  mb = tok - TOK_DIRECT;

  s = get_tok( s, &tok );
  if ( tok != TOK_INT )
    panic_bad_token( tok, "an integer");
  in->b = MODS(tok_int,CORESIZE);


  /*
   * Set flags and ignore the rest of the line
   */
  in->in = (flags << flPOS) | OP( op, m, ma, mb );
  return ASMLINE_OK;
}




/* NAME
 *     asm_file, asm_fname -- assemble a FILE into a warrior
 * 
 * SYNOPSIS
 *     void asm_file( FILE *F, warrior_t *w, unsigned int CORESIZE );
 *     void asm_fname( const char *filename, warrior_t *w,
 *    	               unsigned int CORESIZE );
 * 
 * INPUTS
 *     w        -- warrior_t to assemble into.
 *     F        -- stream to read warrior source from
 *     filename -- path to source file.  May be '-'
 *		   which is interpreted as stdin.
 *     CORESIZE -- just that
 * 
 * DESCRIPTION
 *     These functions assemble a source file into a
 *     warrior_t setting all the non-info fields.
 * 
 * RESULTS
 *    If the warrior assembled correctly, then warrior_t
 *    contains its code and starting offset.  If an error
 *    occured during assembly, an error message is issued
 *    and the program exit()s.
 * 
 * GLOBALS
 *     none as such, subroutines use tok_buf[], tok_int, str_toks[],
 *     MAXLENGTH constant
 * 
 * SEE ALSO
 *     asm_line()
 * 
 * BUGS
 *     Its not really acceptable to exit() on an assembly error.
 */
void
asm_file( FILE *F, warrior_t *w, unsigned int CORESIZE )
{
  char line[MAX_ALL_CHARS];
  insn_t *c;
  int ret;			/* return code from asm_line() */

  w->len = w->start = 0;
  w->have_pin = 0;
  w->pin = 0;
  c = w->code;

  while ( fgets(line, MAX_ALL_CHARS, F) ) {
    ret = asm_line( line, c, CORESIZE );
    if ( ret == ASMLINE_DONE ) break;

    switch ( ret ) {
    case ASMLINE_OK:
      if ( get_flags( c->in ) & fl_START ) {
	w->start = w->len;
	clr_flags( c->in, fl_START );
      }
      if ( w->len < MAXLENGTH) c++; 
      w->len++;
      break;

    case ASMLINE_ORG:
      w->start =  c->a;		/* was `ORG int', get the starting address */
      break;

    case ASMLINE_NONE:
      break;			/* nop */

    case ASMLINE_PIN:
      w->have_pin = 1;
      w->pin = c->a;		/* save PIN. */
      break;

    default:
      fprintf(stderr,"asm.c/asm_file(): illegal return code from asm_line()\n");
      exit(1);
    }
    if ( w->len > MAXLENGTH ) {
      fprintf(stderr, "too many instructions in warrior %d\n", w->no);
      exit(1);
    }
  }
  if ( w->start >= w->len ) {
    fprintf(stderr, "starting address must be inside warrior body\n" );
    exit(1);
  }
}


void
asm_fname( const char *fname, warrior_t *w, unsigned int CORESIZE )
{
  FILE *F;
  int is_stdin = 0;
 
  if ( strcmp( fname, "-" ) == 0 ) {
    F = stdin;
    is_stdin = 1;
  } 
  else
    if (!( F = fopen(fname, "r") )) {
      fprintf(stderr, "can't open file %s\n", fname);
      exit(1);
    }

  asm_file(F, w, CORESIZE);

  if ( !is_stdin ) fclose(F);
}



/* NAME
 *     dis1 -- disasemble an instruction
 *     discore -- disasemble a segment of core
 * 
 * SYNOPSIS
 *     void dis1( char *s, inst_t in, unsigned int CORESIZE );
 *     void discore( inst_t *core, unsigned int start, unsigned int end,
 * 		     unsigned int CORESIZE );
 * 
 * INPUTS
 *     s -- string to print disassembled instruction to. A string
 *          of length 60 should be more than sufficient.
 *     in -- instruction to disassemble
 *     core -- pointer to start of core
 *     start -- core segment start offset
 *     end -- core segment end offset (excluded)
 * 
 * RESULTS
 *     dis1 -- The disassembled instruction is printed to `s'.
 *     discore -- A segment of core is dissasembled and printed
 *                to stdout with core addresses.
 */

void
dis1(char *buf, insn_t in, unsigned int CORESIZE)
{
  int x;
  char *op_s, *mo_s, *ma_s, *mb_s;
  int af, bf;

  x = (in.in >> opPOS) & opMASK;
  switch( x ) {
  case DAT: op_s = "dat"; break;
  case SPL: op_s = "spl"; break;
  case MOV: op_s = "mov"; break;
  case JMP: op_s = "jmp"; break;
  case JMZ: op_s = "jmz"; break;
  case JMN: op_s = "jmn"; break;
  case ADD: op_s = "add"; break;
  case SUB: op_s = "sub"; break;
  case SEQ: op_s = "seq"; break;
  case SNE: op_s = "sne"; break;
  case MUL: op_s = "mul"; break;
  case DIV: op_s = "div"; break;
  case DJN: op_s = "djn"; break;
  case SLT: op_s = "slt"; break;
  case MODM: op_s = "mod"; break;
  case NOP: op_s = "nop"; break;
  case LDP: op_s = "ldp"; break;
  case STP: op_s = "stp"; break;
  default:
    op_s = "???";
  }

  x = (in.in >> moPOS) & moMASK;
  switch ( x ) {
  case mF:  mo_s = "f "; break;
  case mA:  mo_s = "a "; break;
  case mB:  mo_s = "b "; break;
  case mAB: mo_s = "ab"; break;
  case mBA: mo_s = "ba"; break;
  case mX:  mo_s = "x "; break;
  case mI:  mo_s = "i "; break;
  default:
    mo_s = "?";
  }


  x = (in.in >> maPOS) & mMASK;
  switch (x) {
  case DIRECT: ma_s = "$"; break;
  case IMMEDIATE: ma_s = "#"; break;
  case AINDIRECT: ma_s = "*"; break;
  case BINDIRECT: ma_s = "@"; break;
  case APREDEC: ma_s = "{"; break;
  case APOSTINC: ma_s = "}"; break;
  case BPREDEC: ma_s = "<"; break;
  case BPOSTINC: ma_s = ">"; break;
  default: ma_s = "?";
  }

  x = (in.in >> mbPOS) & mMASK;
  switch (x) {
  case DIRECT:    mb_s = "$"; break;
  case IMMEDIATE: mb_s = "#"; break;
  case AINDIRECT: mb_s = "*"; break;
  case BINDIRECT: mb_s = "@"; break;
  case APREDEC:   mb_s = "{"; break;
  case APOSTINC:  mb_s = "}"; break;
  case BPREDEC:   mb_s = "<"; break;
  case BPOSTINC:  mb_s = ">"; break;
  default: mb_s = "?";
  }

  af = in.a <= CORESIZE/2 ? in.a : in.a - CORESIZE;
  bf = in.b <= CORESIZE/2 ? in.b : in.b - CORESIZE;

  sprintf(buf,"%s.%s %s%5d , %s%5d", op_s, mo_s, ma_s, af, mb_s, bf);
}


void
discore( const insn_t *core,
	 int start,
	 int end,
	 unsigned int CORESIZE )
{
  int adr;
  char line[MAX_ALL_CHARS];
  for ( adr = start; adr < end; adr++ ) {
    int i = MODS(adr, CORESIZE);
    dis1( line, core[i], CORESIZE );
    printf("%4d    %s\n", adr, line);
  }
}
